Fix OO token teleporting issue when laying green OO tile

michaeljb opened the pull request:
This breaks a couple connection tests, but I don't entirely understand what's happening in those tests; added some questions in comments in aa40d57.

Still need to make green OO tiles upgradable to brown, so I'm marking this as a Draft PR. Additionally I now seem to be unable to run a route through both cities on a OO tile.

In this flie, it is C&A's turn, they cannot upgrade J4 or run through it twice, though both should be possible.
[cannot_run_or_upgrade_oo.json.txt](https://github.com/tobymao/18xx/files/4654826/cannot_run_or_upgrade_oo.json.txt)

tobymao commented on lib/engine/part/city.rb:
```diff
@@ -73,6 +74,11 @@ def place_token(corporation)
+
+      # returns Array[Integer] of all edges that this city are connected to
+      def connected_edges
+        @connected_edges[tile.rotation] ||= tile.paths.select { |p| p.city == self }.flat_map(&:edges).map(&:num).sort
```
you should never call edges, just call exits

michaeljb replied:
d9d46a6

tobymao commented on lib/engine/part/city.rb:
```diff
@@ -17,6 +17,7 @@ def initialize(revenue, slots = 1)
         @slots = slots.to_i
         @tokens = Array.new(@slots)
         @reservations = []
+        @connected_edges = {}
```
intsead of doing this, let's add to part base an attr called paths

this way every part has a back reference to it's paths

you may need to grep the code a bit to refactor this

michaeljb replied:
Given that when the parts are created, all the paths are not necessarily created (https://github.com/tobymao/18xx/blob/3d43922ace1febb16591b609272a3cd2ea36d036/lib/engine/tile.rb#L68-L104), would you be fine with the different part classes having a `paths` method that calculates `@paths` based on `@tile`?

michaeljb replied:
For example, here's how `Engine::Part::City` could look:

```
def initialize
  # ...
  @paths = {}
end

# ...

def paths
  @paths[tile.rotation] ||= tile.paths.select { |p| p.city == self }
end

def connected_edges
  paths.flat_map(&:exits).sort
end

tobymao commented on lib/engine/hex.rb:
```diff
@@ -60,16 +60,46 @@ def name
+              old_edges & new_edges == old_edges
+            end.compact
+
+            unless cities.one?
```
it it possible that it maps to another city?

michaeljb replied:
Probably not...got rid of this check in cc2a5d3

michaeljb commented:
@tobymao I fixed the tests and removed my comments, but I haven't gone through the connection code very much yet, so I'm not fully confident my "fix" makes sense - 8a0ff24

tobymao commented:
hmmm let me think about it, let's chat on slack m aybe

On Wed, May 20, 2020 at 7:26 PM Michael Brandt <notifications@github.com>
wrote:

> *@michaeljb* commented on this pull request.
> ------------------------------
>
> In lib/engine/part/city.rb
> <https://github.com/tobymao/18xx/pull/418#discussion_r428410631>:
>
> > @@ -17,6 +17,7 @@ def initialize(revenue, slots = 1)
>          @slots = slots.to_i
>          @tokens = Array.new(@slots)
>          @reservations = []
> +        @connected_edges = {}
>
> Given that when the parts are created, all the paths are not necessarily
> created (
> https://github.com/tobymao/18xx/blob/3d43922ace1febb16591b609272a3cd2ea36d036/lib/engine/tile.rb#L68-L104),
> would you be fine with the different part classes having a paths method
> that calculates @paths based on @tile?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tobymao/18xx/pull/418#discussion_r428410631>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB6TF2UNDGY5R2IGP2X5RJ3RSSGLZANCNFSM4NFTOQIQ>
> .
>

tobymao commented:
@michaeljb it makes sense, this error was due to kelsin migrating rotations and i didn't fix the test!

michaeljb commented:
> @michaeljb it makes sense, this error was due to kelsin migrating rotations and i didn't fix the test!

Perfect, that makes way more sense!
